source/compiler/qsc_qasm/src/semantic/lowerer.rs

```diff
@@ -1801,13 +1824,93 @@ impl Lowerer {
+            self.push_const_eval_error(sizeof_invalid_args_error(expr.span, &inputs));
+            return err_expr!(Type::Err, expr.span);
+        };
+
+        // The behavior of `sizeof` changes depending on the type of the first argument, the array.
+        match &first_arg.ty {
+            // If the first argument is an array  or an static reference. We can compute the length
+            // of the requested dimension at lowering time, and the ouput is of type `const uint`.
+            Type::Array(ArrayType { base_ty: _, dims })
+            | Type::StaticArrayRef(StaticArrayRefType {
+                base_ty: _,
+                dims,
+                is_mutable: _,
+            }) => {
+                let second_arg = second_arg as usize;
+                let dims_vec: Vec<_> = dims.clone().into_iter().collect();
+
+                if second_arg >= dims_vec.len() {
+                    self.push_const_eval_error(ConstEvalError::SizeofInvalidDimension(
+                        second_arg,
+                        dims_vec.len(),
+                        expr.span,
+                    ));
+                    return err_expr!(Type::Err, expr.span);
+                }
+
+                Expr::uint(i64::from(dims_vec[second_arg]), expr.span)
+            }
+            // If the first argument is a dynamic reference. We can only compute the length
+            // of the requested dimension at runtime, and the ouput is of type `uint`.
+            Type::DynArrayRef(array_ref) => {
+                if second_arg >= array_ref.num_dims {
+                    self.push_const_eval_error(ConstEvalError::SizeofInvalidDimension(
+                        second_arg as usize,
+                        array_ref.num_dims as usize,
+                        expr.span,
+                    ));
+                    return err_expr!(Type::Err, expr.span);
+                }
+
+                let kind = semantic::ExprKind::SizeofCall(semantic::SizeofCallExpr {
+                    span: expr.span,
+                    fn_name_span: expr.name.span,
+                    array: first_arg,
+                    dim: second_arg,
+                });
+
+                Expr::new(expr.span, kind, Type::UInt(None, false))
+            }
+            _ => {
+                self.push_const_eval_error(sizeof_invalid_args_error(expr.span, &inputs));
+                err_expr!(Type::Err, expr.span)
+            }
+        }
+    }
+
     fn lower_function_call_expr(&mut self, expr: &syntax::FunctionCall) -> semantic::Expr {
         // 1. If the name refers to a builtin function, we defer
         //    the lowering to `lower_builtin_function_call_expr`.
         if let Ok(builtin_fn) = BuiltinFunction::from_str(&expr.name.name) {
             return self.lower_builtin_function_call_expr(expr, builtin_fn);
         }
 
+        if &*expr.name.name == "sizeof" {
+            return self.lower_sizeof_call_expr(expr);
+        }
```

Comment:
Should this be part of `BuiltinFunction::from_str(&expr.name.name)` above and call `lower_sizeof_call_expr` from there?